### PR TITLE
Add "prepare" and "check" watchers

### PR DESCRIPTION
**Pull request by mergeconflict**

### CODE REVIEW

File: evwatcher.c [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+static inline struct event_watcher *
+event_watcher_new(struct event_base *base, event_watcher_cb callback, unsigned type)
+{
+	struct event_watcher *watcher = malloc(sizeof(struct event_watcher));
+	watcher->base = base;
+	watcher->type = type;
+	watcher->callback = callback;
+	TAILQ_INSERT_TAIL(&base->event_watchers[type], watcher, next);
+	return (watcher);
```
Comment (htuch) [inline evwatcher.c]: Nit: I do see plenty of `return 0;` style code in `event.c`, I'd prefer to not use these superfluous parens if we can skip.

File: evwatcher.c [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+struct event_watcher *
+event_watcher_prepare_new(struct event_base *base, event_watcher_cb callback)
+{
+	return (event_watcher_new(base, callback, EVENT_WATCHER_PREPARE_TYPE));
```
Comment (htuch) [inline evwatcher.c]: Ditto etc.

File: include/event2/event_watcher.h [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    https://metacpan.org/pod/distribution/EV/libev/ev.pod#ev_prepare-and-ev_check-customise-your-event-loop!
+    http://docs.libuv.org/en/v1.x/prepare.html, http://docs.libuv.org/en/v1.x/check.html
+
+  Note that we could get away with just passing the watcher, not the base, since event_watcher
+  holds a pointer to its base (see definition in event-internal.h). We'd just need to provide a
+  function like:
+
+    struct event_base *event_watcher_base(struct event_watcher *)
+
+  which is either a convenience or an inconvenience depending how you look at it...
```
Comment (htuch) [inline include/event2/event_watcher.h]: I'd let the work on epoll delay/duration drive this. If libuv keeps it simple, seems fine to go that way.

File: include/event2/event_watcher.h [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+ */
+typedef void (*event_watcher_cb)(struct event_base *, struct event_watcher *);
+
+/**
+  Register a new "prepare" watcher, to be called in the event loop prior to polling for events.
```
Comment (htuch) [inline include/event2/event_watcher.h]: Do we also need to update http://www.wangafu.net/~nickm/libevent-book/ with this? That's basically the libevent docs for most folks. What about release history version files?

File: evwatcher.c [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "event2/event_watcher.h"
+#include "event-internal.h"
+
+static inline struct event_watcher *
+event_watcher_new(struct event_base *base, event_watcher_cb callback, unsigned type)
+{
+	struct event_watcher *watcher = malloc(sizeof(struct event_watcher));
```
Comment (htuch) [inline evwatcher.c]: Should we be using the `mm_alloc` family of functions here? What about possibly `NULL` return? See for example `evbuffer_chain_new` and its handling.

File: test/test-prepare-check.c [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+  /* free the second prepare and check watchers to verify they aren't invoked a second time */
+  event_watcher_free(prepare_callback_2_watcher);
+  event_watcher_free(check_callback_2_watcher);
+  ++timeout_callback_count;
+}
+
+int
+main(int argc, char **argv)
+{
+	base = event_base_new();
```
Comment (htuch) [inline test/test-prepare-check.c]: Nit: weird indent. In fact, the indent in the the test file is generally shorter than the implementation..

File: test/test-prepare-check.c [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+  event_base_dispatch(base);
+  assert(prepare_callback_1_count == 2);
+  assert(prepare_callback_2_count == 1);
+  assert(check_callback_1_count == 2);
+  assert(check_callback_2_count == 1);
+  assert(timeout_callback_count == 1);
+
+  /* this will free the watchers that weren't already freed in timeout_callback */
+  event_base_free(base);
+	return (0);
```
Comment (htuch) [inline test/test-prepare-check.c:152]: More weird indent..

**Review by htuch — CHANGES_REQUESTED** [review]
This looks great overall, ready to ship to libevent GH with some minor comments.

File: event.c [hunk 8]
```diff
@@ -1977,6 +1998,11 @@ event_base_loop(struct event_base *base, int flags)
 			evutil_timerclear(&tv);
 		}
 
+		/* Invoke check watchers after polling for events, and before processing them */
```
Comment (mergeconflict) [inline event.c]: This is definitely in the wrong place. It should be further down, after `evsel->dispatch(...)`

File: event.c [hunk 9]
```diff
@@ -1966,6 +1982,11 @@ event_base_loop(struct event_base *base, int flags)
 			break;
 		}
 
+		/* Invoke prepare watchers before polling for events */
```
Comment (mergeconflict) [inline event.c]: Not sure if this should be here, or just before `evsel->dispatch(...)`. TBD.

File: include/event2/event_watcher.h (hunk 3 above)
Comment (mergeconflict) [reply include/event2/event_watcher.h]: I've added a new `event_watcher_cb_info` struct containing timing information. I'm pretty sure it's actually necessary; there isn't any other good way of determining the timeout value computed by `timeout_next` in `event_base_loop` from outside.

File: include/event2/event_watcher.h (hunk 4 above)
Comment (mergeconflict) [reply include/event2/event_watcher.h]: I'm not sure about the book - it's not in the same repository. And there is a changelog, but it seems to have bitrotted... the last entry is from January 2017. We can get guidance from the libevent maintainers.

File: event.c [hunk 10]
```diff
@@ -1926,6 +1942,8 @@ event_base_loop(struct event_base *base, int flags)
 	struct timeval tv;
 	struct timeval *tv_p;
 	int res, done, retval = 0;
+	struct event_watcher_cb_info watcher_cb_info;
```
Comment (htuch) [inline event.c]: Do you want separate CB info for prepare/check?

File: test/test-prepare-check.c [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+	assert(end_time.tv_sec >= start_time.tv_sec + 1);
+}
+
+int
+main(int argc, char **argv)
+{
+	struct event_base *base = event_base_new();
+
+	/* install prepare and check watchers */
+	event_watcher_prepare_new(base, &prepare_callback_1);
```
Comment (htuch) [inline test/test-prepare-check.c:128]: These tests catch the regression you spotted around check/dispatch ordering, I'm guessing, can you call that out in a comment somewhere?

**Review by htuch — CHANGES_REQUESTED** [review]
LGTM modulo comments.

File: event.c (hunk 10 above)
Comment (mergeconflict) [reply event.c]: I didn't see an immediate need for that, but on second thought, the whole purpose of struct-ifying these is for future flexibility... So yes, yes I want that. Thanks!

File: test/test-prepare-check.c (hunk 11 above)
Comment (mergeconflict) [reply test/test-prepare-check.c:128]: Yeah, I should write a summary at the top... there's a fair bit going on here.

**Pull request closed without merging**